Fix suport for NVVM from conda on Windows

leofang opened the pull request:
## Description

closes #453.

xref: https://github.com/NVIDIA/nvmath-python/blob/02d18a3b88c91432b9a4beef48c116c10121166c/nvmath/_utils.py#L177-L183 

## Checklist
<!-- TODO: - [ ] I am familiar with the [Contributing Guidelines](). -->
- [ ] New or existing tests cover these changes.
- [ ] The documentation is up to date with these changes.

leofang commented:
@rwgk do you still have your Windows instance alive? If so could you double check this locally (since we don't test against conda yet, #280)?

rwgk commented:
> @rwgk do you still have your Windows instance alive? If so could you double check this locally (since we don't test against conda yet, #280)?

I got myself a new instance (for 8 weeks, as you showed me) but only got as far as installing the CUDA driver. (And conda, but I haven't figured out yet how to activate it.)

This is a good motivation for me to continue working on the setup, I'll let you know by tonight (PT) how far I got.

leofang commented:
/ok to test [a14d95e](https://github.com/NVIDIA/cuda-python/pull/563/commits/a14d95eb9bedb6472e85be37d8d9cbd04192bb11)

rwgk commented:
I see one of the tests here is failing, but I downloaded the wheel anyway to my Windows machine.

I used this command (miniforge3):
```
conda create -n ctk128 -c nvidia -c conda-forge python=3.12 cuda-toolkit=12.8.1
```

This zip file:

* https://github.com/NVIDIA/cuda-python/actions/runs/14584592014/artifacts/2982569144

After extracting the .whl file:

```
(ctk128) PS C:\Users\rgrossekunst\Downloads> pip install cuda_bindings-12.8.0-cp312-cp312-win_amd64.whl
Processing c:\users\rgrossekunst\downloads\cuda_bindings-12.8.0-cp312-cp312-win_amd64.whl
Collecting pywin32 (from cuda-bindings==12.8.0)
  Downloading pywin32-310-cp312-cp312-win_amd64.whl.metadata (9.4 kB)
Downloading pywin32-310-cp312-cp312-win_amd64.whl (9.5 MB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 9.5/9.5 MB 59.2 MB/s eta 0:00:00
Installing collected packages: pywin32, cuda-bindings
Successfully installed cuda-bindings-12.8.0 pywin32-310
(ctk128) PS C:\Users\rgrossekunst\Downloads> pip freeze
cuda-bindings @ file:///C:/Users/rgrossekunst/Downloads/cuda_bindings-12.8.0-cp312-cp312-win_amd64.whl#sha256=134d06689b07782f0fde60aa23f35b69750b77a7c073853195d28f1572b3eff8
pywin32==310
setuptools==79.0.0
wheel==0.45.1
(ctk128) PS C:\Users\rgrossekunst\Downloads> python
Python 3.12.10 | packaged by conda-forge | (main, Apr 10 2025, 22:08:16) [MSC v.1943 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> from cuda.bindings import nvvm
>>> nvvm.version()
(2, 0)
>>>
```

Looks like it works?

rwgk commented:
Negative test under very similar conditions:

```
conda create -n ctk128nodev -c nvidia -c conda-forge python=3.12 cuda-toolkit=12.8.1 
```

Using wheel from current `main`:

* https://github.com/NVIDIA/cuda-python/actions/runs/14580338756

* https://github.com/NVIDIA/cuda-python/actions/runs/14580338756/artifacts/2981237073

```
(ctk128nodev) PS C:\Users\rgrossekunst\Downloads\older\cuda-bindings-python312-cuda12.8.0-win-64-d425a8895bd778fffb993263dbf5d9bc631fea22> pip install .\cuda_bindings-12.8.0-cp312-cp312-win_amd64.whl
Processing c:\users\rgrossekunst\downloads\older\cuda-bindings-python312-cuda12.8.0-win-64-d425a8895bd778fffb993263dbf5d9bc631fea22\cuda_bindings-12.8.0-cp312-cp312-win_amd64.whl
Requirement already satisfied: pywin32 in c:\programdata\miniforge3\envs\ctk128nodev\lib\site-packages (from cuda-bindings==12.8.0) (310)
Installing collected packages: cuda-bindings
Successfully installed cuda-bindings-12.8.0
(ctk128nodev) PS C:\Users\rgrossekunst\Downloads\older\cuda-bindings-python312-cuda12.8.0-win-64-d425a8895bd778fffb993263dbf5d9bc631fea22> python
Python 3.12.10 | packaged by conda-forge | (main, Apr 10 2025, 22:08:16) [MSC v.1943 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> from cuda.bindings import nvvm
>>> nvvm.version()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "cuda\\bindings\\nvvm.pyx", line 76, in cuda.bindings.nvvm.version
  File "cuda\\bindings\\nvvm.pyx", line 90, in cuda.bindings.nvvm.version
  File "cuda\\bindings\\cynvvm.pyx", line 15, in cuda.bindings.cynvvm.nvvmVersion
  File "cuda\\bindings\\_internal\\nvvm.pyx", line 258, in cuda.bindings._internal.nvvm._nvvmVersion
  File "cuda\\bindings\\_internal\\nvvm.pyx", line 115, in cuda.bindings._internal.nvvm._check_or_init_nvvm
  File "cuda\\bindings\\_internal\\nvvm.pyx", line 87, in cuda.bindings._internal.nvvm.load_library
RuntimeError: Failed to load nvvm
>>> 
```

I think in combination with the previous comment, that's conclusive.

rwgk commented:
The path in this PR is definitely correct:
```
(ctk128) PS C:\Users\rgrossekunst\Downloads> dir C:\ProgramData\miniforge3\envs\ctk128\Library\nvvm\bin

    Directory: C:\ProgramData\miniforge3\envs\ctk128\Library\nvvm\bin

Mode                 LastWriteTime         Length Name
----                 -------------         ------ ----
-a----         2/21/2025   9:18 PM       67792896 cicc.exe
-a----         2/21/2025   9:18 PM       52873216 nvvm64_40_0.dll

(ctk128) PS C:\Users\rgrossekunst\Downloads> dir C:\ProgramData\miniforge3\envs\ctk128nodev\Library\nvvm\bin

    Directory: C:\ProgramData\miniforge3\envs\ctk128nodev\Library\nvvm\bin

Mode                 LastWriteTime         Length Name
----                 -------------         ------ ----
-a----         2/21/2025   9:18 PM       67792896 cicc.exe
-a----         2/21/2025   9:18 PM       52873216 nvvm64_40_0.dll
```

rwgk commented:
I didn't push this, to not conflict with your work:

```diff
commit ceb62f1db9f07187c859f202b02a1e9910ae296f (HEAD -> fix_win_conda_nvvm)
Author: Ralf W. Grosse-Kunstleve <rgrossekunst@nvidia.com>
Date:   Mon Apr 21 22:03:58 2025 -0700

    Fix two bugs: 1. "conda" needs to be skipped if CONDA_PREFIX is not defined (that is a new bug). 2. Existing indentation error (spotted by ChatGPT).

diff --git a/cuda_bindings/cuda/bindings/_internal/nvvm_windows.pyx b/cuda_bindings/cuda/bindings/_internal/nvvm_windows.pyx
index 9243bf07..44ec16f4 100644
--- a/cuda_bindings/cuda/bindings/_internal/nvvm_windows.pyx
+++ b/cuda_bindings/cuda/bindings/_internal/nvvm_windows.pyx
@@ -62,23 +62,26 @@ cdef load_library(const int driver_ver):
 
         # Next, check if DLLs are installed via pip or conda
         for sp in get_site_packages():
-            if sp == "conda" and "CONDA_PREFIX" in os.environ:
+            if sp == "conda":
                 # nvvm is not under $CONDA_PREFIX/lib, so it's not in the default search path
-                mod_path = os.path.join(os.environ["CONDA_PREFIX"], "Library", "nvvm", "bin")
+                conda_prefix = os.environ.get("CONDA_PREFIX")
+                if conda_prefix is None:
+                    continue
+                mod_path = os.path.join(conda_prefix, "Library", "nvvm", "bin")
             else:
                 mod_path = os.path.join(sp, "nvidia", "cuda_nvcc", "nvvm", "bin")
             if not os.path.isdir(mod_path):
                 continue
             os.add_dll_directory(mod_path)
-        try:
-            handle = win32api.LoadLibraryEx(
-                # Note: LOAD_LIBRARY_SEARCH_DLL_LOAD_DIR needs an abs path...
-                os.path.join(mod_path, dll_name),
-                0, LOAD_LIBRARY_SEARCH_DEFAULT_DIRS | LOAD_LIBRARY_SEARCH_DLL_LOAD_DIR)
-        except:
-            pass
-        else:
-            break
+            try:
+                handle = win32api.LoadLibraryEx(
+                    # Note: LOAD_LIBRARY_SEARCH_DLL_LOAD_DIR needs an abs path...
+                    os.path.join(mod_path, dll_name),
+                    0, LOAD_LIBRARY_SEARCH_DEFAULT_DIRS | LOAD_LIBRARY_SEARCH_DLL_LOAD_DIR)
+            except:
+                pass
+            else:
+                break
 
         # Finally, try default search
         try:
```

leofang commented:
@rwgk Yes. I fixed conda, but then broke wheels 😓 It turns out that I actually introduced a pedantic bug in all DLL loading logic (to cuda-bindings and in nvmath-python) that's not discovered until we apply it to NVVM... Both should be fixed now with commit https://github.com/NVIDIA/cuda-python/commit/21a7fb4a8dfbf71145b3389cf1d7cab0a252ec67.

We keep looping over all possible `mod_path` but we did not stop when hitting a valid one. Then, by the time we use it to assemble an absolute path to the DLL, it might not be a valid path. This is exposed because we added conda to the end of the loop only for NVVM, and the previous found (valid) wheel path got messed up. It's now fixed for all places where we do DLL loading. I think it's a bug unlikely to hit because without conda we only search among user/global site-packages, and it's unlikely. though not impossible, to find multiple copies of CUDA wheels there.

leofang commented:
> I didn't push this, to not conflict with your work

Feel free to push and I'll retest locally! @rwgk I think your patch is better in that we actually try to load in every loop iteration, instead of trying it after the first hit.

leofang commented:
CI is green now. Let me apply your patch and then we can test/merge.

rwgk commented:
/ok to test a0baf7197b2e4e1e7898751a8a1ced14aa3cbe06

rwgk commented:
To explain commit a0baf7197b2e4e1e7898751a8a1ced14aa3cbe06:

* Common to all cases (nvrtc, nvJitLink, nvvm): Ensure that `mod_path` is used only when it is defined. — Note that the loop over site-packages may have no iterations.

* For nvrtc: The DLL search order is now consistent with that of nvJitLink, nvvm.

kkraus14 commented on cuda_bindings/cuda/bindings/_bindings/cynvrtc.pyx.in:
```diff
@@ -60,36 +60,37 @@ cdef int cuPythonInit() except -1 nogil:
+                            os.path.join(mod_path, "nvrtc64_120_0.dll"),
+                            0, LOAD_LIBRARY_SEARCH_DEFAULT_DIRS | LOAD_LIBRARY_SEARCH_DLL_LOAD_DIR)
+
+                        # Note: nvrtc64_120_0.dll calls into nvrtc-builtins64_*.dll which is
+                        # located in the same mod_path.
+                        # Update PATH environ so that the two dlls can find each other
+                        os.environ["PATH"] = os.pathsep.join((os.environ.get("PATH", ""), mod_path))
```
Should we be manually loading this dll instead of modifying `PATH`? Regardless, we don't need to fix this in this PR.

rwgk replied:
nvrtc is very special, because of the builtins runtime dependency.

In one of my ChatGPT chats about it a few days ago, it suggested strongly to do both, `os.environ["PATH"]` update, and `os.add_dll_directory(mod_path)`. Therefore I'm carrying that into the path_finder.

rwgk replied:
I forgot to add: I somehow have it in my mind that @leofang made a remark that we shouldn't load the `builtins` ourselves. Leo, does that make sense?

leofang replied:
Pre-loading a DLL works (it's [what I did in nvmath](https://github.com/NVIDIA/nvmath-python/blob/02d18a3b88c91432b9a4beef48c116c10121166c/nvmath/_utils.py#L138)). My reason against pre-loading is not because it does not work, but because I am not willing to maintain other libraries' implementation details (dlopen, which has no `DT_NEEDED` or package metadata for us to inspect). This can easily go out-of-date as these libraries evolve.

rwgk replied:
Thanks Leo, I linked your comment here:

https://github.com/NVIDIA/cuda-python/pull/558/commits/1344621cf6b542df18eef8fede39b26b102e6449

We can weigh the pros-and-cons of "soft dependency pre-loading" vs `os.environ["PATH"]` & `os.add_dll_directory()` management when the path_finder code is complete.

rwgk commented:
/ok to test 03e6e4a170906fb144e967701deb22b1c28840fb

leofang commented on cuda_bindings/cuda/bindings/_bindings/cynvrtc.pyx.in:
```diff
@@ -60,36 +60,37 @@ cdef int cuPythonInit() except -1 nogil:
-                pass
+                if os.path.isdir(mod_path):
+                    os.add_dll_directory(mod_path)
+                    LOAD_LIBRARY_SEARCH_DEFAULT_DIRS = 0x00001000
+                    LOAD_LIBRARY_SEARCH_DLL_LOAD_DIR = 0x00000100
```
nit: let's move this outside of the loop

rwgk replied:
Retesting with commit 98d7a91d31f6abc50485ae5c4d2fc26e8858c5e1

leofang commented on cuda_bindings/cuda/bindings/_internal/nvvm_windows.pyx:
```diff
@@ -50,44 +50,42 @@ cdef load_library(const int driver_ver):
         # Finally, try default search
+        # Only reached if DLL wasn't found in any site-package path
         try:
             handle = win32api.LoadLibrary(dll_name)
```
```suggestion
            return win32api.LoadLibrary(dll_name)
```

rwgk replied:
Thanks for catching that!

leofang replied:
So very happy when the CI is doing the job right! We really need to get the conda counterpart in asap... (#280, cc @cryos for vis 👼 )

rwgk commented:
/ok to test 1a05efc8dff734b5105f3c01e0c5f7691172c776

leofang commented:
Thanks, Ralf/Keith!
